File: CodeForces/String Coloring - easy version/sol.cpp
// http://codeforces.com/contest/1296/problem/E1
#include <bits/stdc++.h>
#include <cassert>

typedef unsigned long long ull;
typedef unsigned long ul;

using namespace std;

class Node {
    public:
    char letter;
    int idx;
    bool color = false;
    vector<Node*> dependencies;
    
    Node(char ltr, int idx) {
        this->letter = ltr;
        this->idx = idx;
    }

    bool operator== (const Node &other) {
        return letter == other.letter and idx == other.idx;
    }

    void addDependency(Node *node) {
        assert(!count(dependencies.begin(), dependencies.end(), node));
        dependencies.push_back(node);
        // cout << dependencies.size() << endl;
    }
};

// Checks if a node's dependency path exceeds given max depth
bool checkDepDepth(Node *node, int maxDepth=1) {
    if (maxDepth < 0) return false;
    for(auto n : node->dependencies) {
        n->color = !node->color;
        if (!checkDepDepth(n, maxDepth-1)) return false;
    }
    return true;
}

int main(int argc, const char** argv) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int x;
    cin >> x;
    string chars;
    cin >> chars;
    vector<Node> nodes;
    for(int i = 0; i < chars.length() ; i++) {
        nodes.push_back(Node(chars.at(i), i));
    }
    for (int i1 = 0; i1 < nodes.size(); i1++) {
        Node &n1 = nodes.at(i1);
        for (int i2 = i1 + 1; i2 < nodes.size(); i2++) {
            Node &n2 = nodes.at(i2);
            if (n2.letter < n1.letter) {
                n1.addDependency(&n2);
            }
        }
    }
    string colors;
    for(auto &n : nodes) {
        if (!checkDepDepth(&n, 1)) {
            cout << "NO" << endl;
            return 0;
        }
        colors.push_back(n.color ? '1' : '0');
    }
    cout << "YES" << endl << colors << endl;
}

File: CodeForces/String Coloring - easy version/sol.py
#!/usr/bin/python3
#http://codeforces.com/contest/1296/problem/E1

class Node:
    idx = None
    letter = None
    color = False

    def __init__(self, idx, letter):
        assert idx >= 0 and letter
        self.dependencies = []
        self.dependents = []
        self.idx = idx
        self.letter = letter
    
    def __repr__(self):
        return f'Node<{self.idx, self.letter}>'

    def add_dependency(self, node):
        assert type(node) is Node
        self.dependencies.append(node)
        node.dependents.append(self)

input()
string = input()
nodes = [Node(*x) for x in enumerate(string)]

def check_dep_depth(node, max_depth=1):
    """Check if any node dep path exceeds max_depth"""
    if max_depth < 0:
        return False
    for n in node.dependencies:
        n.color = not node.color
        if not check_dep_depth(n, max_depth-1):
            return False
    return True

for idx, n1 in enumerate(nodes):
    if idx == len(nodes) - 1:
        continue
    for n2 in nodes[idx+1:]:
        if n2.letter < n1.letter:
            # print(n1, n2)
            n1.add_dependency(n2)

for n in nodes:
    if not check_dep_depth(n, 1):
        print('NO')
        exit()

print('YES')
print(''.join(str(int(n.color)) for n in nodes))